src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt

```diff
@@ -63,44 +49,40 @@ class VisitService(
   @Autowired
   private lateinit var visitDtoBuilder: VisitDtoBuilder
 
-  @Autowired
-  private lateinit var applicationService: ApplicationService
-
-  @Autowired
-  private lateinit var sessionTemplateService: SessionTemplateService
-
   companion object {
     val LOG: Logger = LoggerFactory.getLogger(this::class.java)
     const val MAX_RECORDS = 10000
-    const val AMEND_EXPIRED_ERROR_MESSAGE = "Visit with booking reference - %s is in the past, it cannot be %s"
   }
 
   fun bookVisit(applicationReference: String, bookingRequestDto: BookingRequestDto): VisitDto {
-    if (applicationService.isApplicationCompleted(applicationReference)) {
-      LOG.debug("The application $applicationReference has already been booked!")
-      // If already booked then just return object and do nothing more!
-      val visit = visitRepository.findVisitByApplicationReference(applicationReference)!!
-      return visitDtoBuilder.build(visit)
+    val alreadyBookedVisit = visitStoreService.checkBookingAlreadyMade(applicationReference)
+    alreadyBookedVisit?.let {
+      return alreadyBookedVisit
     }
-    // Need to set application complete at earliest opportunity to prevent two bookings from being created, Edge case.
-    applicationService.completeApplication(applicationReference)
-
-    val application = applicationService.getApplicationEntity(applicationReference)
 
-    val existingBooking = visitRepository.findVisitByApplicationReference(application.reference)
+    val booking = visitStoreService.createOrUpdateBooking(applicationReference, bookingRequestDto)
+    return processBookingEvents(booking, bookingRequestDto)
+  }
 
-    // application validity checks
-    applicationValidationService.validateApplication(bookingRequestDto, application, existingBooking)
+  fun updateBookedVisit(applicationReference: String, bookingRequestDto: BookingRequestDto): VisitDto {
+    val alreadyBookedVisit = visitStoreService.checkBookingAlreadyMade(applicationReference)
```

Comment:
is this logic correct given this is an update? won't this always return`alreadyBookedVisit`   for an update?

Reply:
As discussed via slack, leaving this as is. All logic for this refactor will remain, can look to optimise and improve in the future if needed.

Reason for the check, is a pre-existing issue where 2 API calls were made at same time for the same application, it caused issues with the 1st call creating a booking, then the 2nd call overwriting the application for that booking causing issues. The check is a crude way to control this as the Db has no locking mechanism for multi requests at same time.